File: 2369-check-if-there-is-a-valid-partition-for-the-array/2369-check-if-there-is-a-valid-partition-for-the-array.java
class Solution {
    public boolean validPartition(int[] nums) {
        // boolean[] dp = int dp[nums+1];
        HashMap<Integer,Boolean> memo = new HashMap<>();
        return partition(nums,0,memo);
    }
    private boolean partition(int[] nums, int start,HashMap<Integer,Boolean> memo)
    {
        if(start==nums.length) return true;
        
        int curr = start;
        if(memo.containsKey(curr))
        {
            return memo.get(curr);
        }
        
        boolean c1 =false,c2=false,c3=false;
        
        if(start+2<nums.length && nums[start]== nums[start+1] && nums[start+1]==nums[start+2]){
            c1 = partition(nums,start+3,memo);
        }
        if(start+2<nums.length && nums[start+1]== nums[start]+1 && nums[start+2]==nums[start+1]+1){
            c2 = partition(nums,start+3,memo);
        }
        if(start+1<nums.length && nums[start+1]==nums[start]){
            c3 = partition(nums,start+2,memo);
        }
        memo.put(start,c1||c2||c3);
        return c1||c2||c3;
        
    }
}

File: 2369-check-if-there-is-a-valid-partition-for-the-array/2369-check-if-there-is-a-valid-partition-for-the-array.cpp
class Solution
{
    public:
        bool validPartition(vector<int> &nums)
        {
            vector<int> dp(nums.size() + 1, -1);
            return isPartition(0, nums, dp);
        }

    bool isPartition(int start, vector<int> &nums, vector<int> &dp)
    {
        // if (start > nums.size())
        //     return false;
        if (start == nums.size())
            return true;
        if (dp[start] != -1)
            return dp[start];
        bool ans1 = false, ans2 = false, ans3 = false;
        if (start + 2 < nums.size() && nums[start] == nums[start + 1] && nums[start + 1] == nums[start + 2])
            ans1 = isPartition(start + 3, nums, dp);

        if (start + 1 < nums.size() && nums[start] == nums[start + 1])
            ans2 = isPartition(start + 2, nums, dp);

        if (start + 2 < nums.size() && nums[start] == nums[start + 1] - 1 && nums[start + 1] == nums[start + 2] - 1)
            ans3 = isPartition(start + 3, nums, dp);

        return dp[start] = ans1 || ans2 || ans3;
    }
};